third_party/move/move-compiler-v2/src/pipeline/avail_copies_analysis.rs

```diff
@@ -0,0 +1,348 @@
+    stackless_bytecode::{Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Collection of definitely available copies.
+/// For a copy `a := b`, we store the key-value pair `(a, b)` in the internal map.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AvailCopies(BTreeMap</*dst*/ TempIndex, /*src*/ TempIndex>);
+
+impl AvailCopies {
+    /// Create a new (empty) collection of definitely available copies.
+    fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+
+    /// Make a copy `dst := src` available.
+    /// If either `dst` or `src` is in `borrowed_locals`, then the copy is not made available.
+    /// To call this method, `dst := x` should not already be available for any `x`.
+    fn make_copy_available(
+        &mut self,
+        dst: TempIndex,
+        src: TempIndex,
+        borrowed_locals: &BTreeSet<TempIndex>,
+    ) {
+        if src == dst {
+            // No need to make a copy available for self-assignments.
+            return;
+        }
+        // Note that we are conservative here for the sake of simplicity, and disallow copies
+        // when either `dst` or `src` is borrowed. We could track more copies as available by using
+        // reference analysis.
+        if !borrowed_locals.contains(&dst) && !borrowed_locals.contains(&src) {
+            let old_src = self.0.insert(dst, src);
+            if let Some(old_src) = old_src {
+                panic!(
+                    "ICE: copy `$t{} = $t{}` already available, \
+                        cannot have `$t{} = $t{}` available as well",
+                    dst, old_src, dst, src
+                );
+            }
+        }
+    }
+
+    /// Kill all available copies of the form `x := y` where `x` or `y` is `tmp`.
+    /// If `tmp` is in `borrowed_locals`, then no copies are killed.
+    fn kill_copies_with(&mut self, tmp: TempIndex, borrowed_locals: &BTreeSet<TempIndex>) {
+        if !borrowed_locals.contains(&tmp) {
+            // TODO: consider optimizing the following operation by keeping a two-way map between
+            // `dst -> src` and `src -> set(dst)`. Another optimization to consider is to use im::OrdMap.
+            self.0.retain(|dst, src| *dst != tmp && *src != tmp);
+        }
+    }
+
+    /// Given a set of available copies: `tmp_1 := tmp_0, tmp_2 := tmp_1,..., tmp_n := tmp_n-1`, forming
+    /// the copy chain: `tmp_0 --copied-to--> tmp_1 --copied-to--> tmp_2 -> ... -> tmp_n-1 -> tmp_n`,
+    /// return the head of the copy chain `tmp_0` for any input `tmp_x` (x in 0..=n) in the chain.
+    ///
+    /// Note that it is a required invariant that the copy chain is acyclic, else we panic.
+    /// The natural way of constructing the copy chain for move bytecode (like in this file) ensures this.
+    pub fn get_head_of_copy_chain(&self, mut tmp: TempIndex) -> TempIndex {
+        let mut visited = BTreeSet::from([tmp]);
```

Comment:
This seems like an expensive way of checking, especially since you expect to never happen.  How about a counter instead?
```
   let mut counter = 0;
   let usize max_counter = self.0.len();
   while let Some(src) = self.0.get(&tmp) {
      counter += 1;
      if counter >= max_counter {  panic ... }
      tmp = *src;
   }
   tmp

```

Reply:
Nice idea.

What do you think about removing this checking related code completely? I added it initially as a "debug" code to make sure I did not have bugs in the way I construct copy chains, but I think based on how they are constructed, we can prove that there will be no cycles.

Reply:
I think we should keep this kind of checks, they are also good for documentation. As to serve as doc/spec, the current code is a better readable than Brian's... Performance probably not that important, these chains are not that deep. But both are good with me.